File: 0031/solution.cpp
#include <gtest/gtest.h>

#include <cstdlib>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
 public:
  void nextPermutation(vector<int>& nums) {
    if (nums.size() == 1) return;
    int last_idx = -1;
    for (int i = 0; i < nums.size() - 1; i++) {
      if (nums[i] < nums[i+1]) {
        last_idx = i;
      }
    }
    if (last_idx == (nums.size() - 1) - 1) {
      swap(nums, last_idx, last_idx + 1);
    } else if (last_idx == -1) {
      reverse(nums, 0);
    } else {
      // 2,1,5,4,3 -> 2,3,1,4,5
      // 2,3,5,4,1 -> 2,4,1,3,5
      // 1,5,1 -> 5,1,1
      int idx_to_swap = last_idx + 1;
      for (int i = last_idx + 1; i < nums.size(); i++) {
        if (nums[last_idx] >= nums[i]) break;
        idx_to_swap = i;
      }
      // first, swap 3 and 4
      // 2,3,5,4,1 -> 2,4,5,3,1
      swap(nums, last_idx, idx_to_swap);
      reverse(nums, last_idx + 1);
    }
  }

  void swap(vector<int>& nums, int i, int j) {
    int tmp = nums[i];
    nums[i] = nums[j];
    nums[j] = tmp;
  }

  void reverse(vector<int>&nums, int i) {
    int start_idx = i;
    int end_idx = nums.size() - 1;
    while (start_idx < end_idx) {
      swap(nums, start_idx, end_idx);
      start_idx++;
      end_idx--;
    }
  }
};

int main() {
  Solution sol;
  vector<int> d, expected;

  d = {1}, expected = {1};
  sol.nextPermutation(d);
  EXPECT_EQ(expected, d);

  d = {1, 2}, expected = {2, 1};
  sol.nextPermutation(d);
  EXPECT_EQ(expected, d);

  d = {1, 2, 3}, expected = {1, 3, 2};
  sol.nextPermutation(d);
  EXPECT_EQ(expected, d);

  d = {1, 3, 2}, expected = {2, 1, 3};
  sol.nextPermutation(d);
  EXPECT_EQ(expected, d);

  d = {2, 1, 3}, expected = {2, 3, 1};
  sol.nextPermutation(d);
  EXPECT_EQ(expected, d);

  d = {2, 3, 1}, expected = {3, 1, 2};
  sol.nextPermutation(d);
  EXPECT_EQ(expected, d);

  d = {3, 1, 2}, expected = {3, 2, 1};
  sol.nextPermutation(d);
  EXPECT_EQ(expected, d);

  d = {3, 2, 1}, expected = {1, 2, 3};
  sol.nextPermutation(d);
  EXPECT_EQ(expected, d);

  d = {1, 1, 5}, expected = {1, 5, 1};
  sol.nextPermutation(d);
  EXPECT_EQ(expected, d);

  d = {1, 5, 1}, expected = {5, 1, 1};
  sol.nextPermutation(d);
  EXPECT_EQ(expected, d);

  d = {5, 1, 1}, expected = {1, 1, 5};
  sol.nextPermutation(d);
  EXPECT_EQ(expected, d);

  d = {1, 2, 3, 4}, expected = {1, 2, 4, 3};
  sol.nextPermutation(d);
  EXPECT_EQ(expected, d);

  d = {1, 2, 4, 3}, expected = {1, 3, 2, 4};
  sol.nextPermutation(d);
  EXPECT_EQ(expected, d);

  d = {4, 3, 2, 1}, expected = {1, 2, 3, 4};
  sol.nextPermutation(d);
  EXPECT_EQ(expected, d);

  exit(EXIT_SUCCESS);
}

File: 0031/solution.rs
struct Solution {}

impl Solution {
    pub fn next_permutation(nums: &mut Vec<i32>) {
        // 1 4 5 2 3 -> 1 4 5 3 2
        // 1 4 5 3 2 -> 1 5 2 3 4
        // index whose value is bigger than the right one
        let mut last_idx = 0;
        for idx in 1..nums.len() {
            if nums[idx-1] < nums[idx] {
                last_idx = idx;
            }
        }
        // 5 4 3 2 1 -> 1 2 3 4 5
        if (last_idx == 0) {
            Solution::reverse(nums, 0);
            return;
        }
        /*
        if last_idx == nums.len() - 1 {
            Solution::swap(nums, last_idx - 1, last_idx);
        }
        */
        // 1 4 5 2 3
        // 1 4 5 3 2 -> 1 5 2 3 4
        let idx_to_swap = last_idx - 1;
        let mut second_idx_to_swap = last_idx;
        for idx in last_idx..nums.len()-1 {
            if nums[last_idx] < nums[idx] {
                break;
            }
            second_idx_to_swap = idx;
        }
    }

    pub fn swap(nums: &mut Vec<i32>, i: usize, j: usize) {
        let tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }

    pub fn reverse(nums: &mut Vec<i32>, i: usize) {
        let mut start_idx = i;
        let mut end_idx = nums.len() - 1;
        while start_idx < end_idx {
            Solution::swap(nums, start_idx, end_idx);
            start_idx++;
            end_idx--;
        }
    }
}

fn main() {
    let mut v: Vec<i32> = vec![1, 2, 3, 4, 5];
    Solution::next_permutation(&mut v);
    assert_eq!(v, [1, 2, 3, 5, 4]);

    v = vec![1, 2, 3, 5, 4];
    Solution::next_permutation(&mut v);
    assert_eq!(v, [1, 2, 4, 3, 5]);

    v = vec![1, 2, 4, 5, 3];
    Solution::next_permutation(&mut v);
    assert_eq!(v, [1, 2, 5, 3, 4]);

    v = vec![1, 4, 5, 3, 2];
    Solution::next_permutation(&mut v);
    assert_eq!(v, [1, 5, 2, 3, 4]);

    v = vec![1, 5, 4, 3, 2];
    Solution::next_permutation(&mut v);
    assert_eq!(v, [2, 1, 3, 4, 5]);
}
